pumpkin-lib/src/engine/constraint_satisfaction_solver.rs

```diff
@@ -1375,18 +1401,114 @@ impl ConstraintSatisfactionSolver {
             return PropagationStatusOneStepCP::FixedPoint;
         }
 
-        let num_trail_entries_before = self.assignments_integer.num_trail_entries();
-
+        let cp_trail_length = self.assignments_integer.num_trail_entries();
+        let is_at_root = self.get_decision_level() == 0;
         let propagator_id = self.propagator_queue.pop();
+        let tag = self.cp_propagators.get_tag(propagator_id);
         let propagator = &mut self.cp_propagators[propagator_id];
-        let context = PropagationContextMut::new(
-            &mut self.assignments_integer,
-            &mut self.reason_store,
-            &mut self.assignments_propositional,
-            propagator_id,
-        );
 
-        let result = match propagator.propagate(context) {
+        let propagation_status = {
+            let context = PropagationContextMut::new(
+                &mut self.assignments_integer,
+                &mut self.reason_store,
+                &mut self.assignments_propositional,
+                propagator_id,
+            );
+
+            propagator.propagate(context)
+        };
+
+        if is_at_root && self.internal_parameters.proof_log.is_logging_inferences() {
```

Comment:
It is unclear what is being done here and the method goes from simple and understandable to long and difficult; could be refactored?